README.md
Bunyan
======
![](http://img2.timeinc.net/ew/dynamic/imgs/080612/Paul-Bunyan-Blue-Ox_l.jpg)

Bunyan is a thin ruby wrapper around a MongoDB [capped collection](http://www.mongodb.org/display/DOCS/Capped+Collections), 
created with high-performance, flexible logging in mind.

Bunyan is not intended to be used as a drop-in replacement for the default
rails logger. If you need to do this, please see peburrows' [mongo\_db\_logger](http://github.com/peburrows/mongo_db_logger).

However, there is planned support for a flexible middleware component to drop in
to your rails app. 

Install
=======
    gem install bunyan

Configure
=========
The only configuration options required are the database and collection name.

For rails apps, put the following config block in an initializer.

    # config/initializers/bunyan.rb
    Bunyan::Logger.configure do
      # required options
      database   'bunyan_logger'
      collection 'development_log'

      # optional
      host     'some.remote.host' # defaults to localhost
      port     '12345'            # defaults to 27017
      disabled true

      # other connection option
      # Specify a valid ruby driver connection object (single / pair)
      # if you do so, host / port are ignored 
      connection = $connection
    end
    
    # Gemfile
    gem 'bson_ext', '1.1' # set the version to whatever the current version of bson the mongo gem is using

Usage
=====
You can access the bunyan logger anywhere is your app via Bunyan::Logger.
The Logger class is implemented as a singleton, and you can call any instance
method on the Logger class. This is implemented with some method_missing magic,
which is explained in a bit more detail below.

Middleware
==========
To make use of Bunyan in your web apps, you need some way to capture request data. 
I have added an example middleware (which is extremely, extremely ugly) which you
can find at `examples/middleware.rb`. 

There are plans for a standard middleware API, but it has not yet been implemented.
If you have interest in helping/collaborating on this feature, please get in touch
with me. 

Until the middleware API is complete, please use this example middleware at your
own risk!

Internals
=========
Bunyan makes heavy usage of method\_missing both at the class and instance level.
At the class level, this is purely for convenience to pass missing class methods
to the singleton instance, which precludes us from needing to use Logger.instance
directly. At the instance level, method\_missing is used to provide a very thin
layer around MongoDB, via the Mongo driver.

The net effect of this is that there is no reason to muck around with 
calling methods directly on Bunyan::Logger.instance, because all methods that 
don't already exist at the class level will be sent to Bunyan::Logger.instance.

    Bunyan::Logger.count == Bunyan::Logger.instance.count # => true

Also, you can call any instance methods Bunyan::Logger that you would otherwise 
call on a Mongo collection object.

More
====
* [MongoDB Capped Collections](http://www.mongodb.org/display/DOCS/Capped+Collections)
* [Mongo Ruby Driver](http://github.com/mongodb/mongo-ruby-driver)

TODO
====
* Ability to specify auth creds for making a mongo connection
* <del>Fail silently if no mongo server running</del>
* Ability to limit bunyan to only run in certain environments
* Add middleware client for easy drop-in to rails/rack apps
* <del>Add ability to connect to a remote mongo server</del>
* <del>Ability to configure size of capped collection</del>


CHANGELOG.md
Version 0.5.0
=============
* Added ActiveRecord-style "silent" connection reconnect
* Added #abort_on_failed_reconnect configuration option
* Upped dependent version of mongo gem to ~> 1.2.4

Version 0.4.0
=============
* New configuration syntax (see readme)
* Ability to pass in custom connection object

Version 0.3.0
=============
* Can now specify and connect a remote host and alternate port

Version 0.2.2
=============
* No longer pass messages to the collection if a Mongo connection could not be established
* Bunyan now uses the new bson_ext c extension, per version 0.20.0 of the ruby driver

Version 0.2.1
=============
* Bunyan now fails silently when a connection error occurs

Version 0.2.0
=============
* Moved all configuration-related stuff to the Config class
* Added ability to set the collection size in configuration block
* Bunyan attributes (:connection, :collection, :db) now map directly to mongo counterparts


spec/bunyan_spec.rb
require 'spec_helper'

describe Bunyan::Logger do
  before do
    Mongo::Connection.unstub!(:new)
    @logger = Bunyan::Logger.instance
    @mock_database = nil
  end

  it 'should have a connection' do
    configure_test_db
    @logger.connection.should be_instance_of Mongo::Connection
  end

  it 'should have a reference to the mongo db object' do
    configure_test_db
    @logger.db.should be_instance_of Mongo::DB
  end

  it 'should have a config hash' do
    Bunyan::Logger.config.should respond_to :[]
  end


  it 'should use the mongo c extension' do
    defined?(CBson::VERSION).should_not be_nil
  end
end

describe 'when a mongod instance is not running' do
  context "when initializing the bunyan setup" do
    before do
      Mongo::Connection.stub!(:new).and_raise(Mongo::ConnectionFailure)
    end

    it 'should not blow up' do
      lambda {
        Bunyan::Logger.configure do |c|
          c.database   'doesnt_matter'
          c.collection 'b/c mongod isnt running'
        end
      }.should_not raise_exception(Mongo::ConnectionFailure)
    end

    it 'should mark bunyan as disabled' do
      Bunyan::Logger.configure do |c|
        c.database   'doesnt_matter'
        c.collection 'b/c mongod isnt running'
      end
      Bunyan::Logger.should be_disabled
    end
  end

  context "when sending queries to Bunyan when a connection error occurs" do
    class Mongo::Connection
      attr_accessor :conn_fail

      def new_send_message_on_socket(*args)
        if conn_fail && conn_fail > 0
          self.conn_fail -= 1
          raise Mongo::ConnectionFailure
        else
          old_send_message_on_socket(*args)
        end
      end
    end

    before do
      Mongo::Connection.send(:alias_method, :old_send_message_on_socket, :send_message_on_socket)
      Mongo::Connection.send(:alias_method, :send_message_on_socket, :new_send_message_on_socket)
    end

    after do
      Mongo::Connection.send(:alias_method, :send_message_on_socket, :old_send_message_on_socket)
    end

    before do
      unstub_mongo
      configure_test_db
    end

    it "should not raise an exception" do
      Bunyan::Logger.connection.conn_fail = 1

      lambda {
        Bunyan::Logger.count
      }.should_not raise_error(Mongo::ConnectionFailure)
    end

    context "by default" do
      it "should not raise an exception if a connection error occurs on multiple requests" do
        Bunyan::Logger.connection.conn_fail = 2

        lambda {
          Bunyan::Logger.count
        }.should_not raise_error(Mongo::ConnectionFailure)
      end
    end

    context "when abort_on_failed_reconnect is true" do
      before do
        configure_test_db do |c|
          c.abort_on_failed_reconnect true
        end
      end

      it "should raise an exception when more than one failed request occurs" do
        Bunyan::Logger.connection.conn_fail = 2

        lambda {
          Bunyan::Logger.count
        }.should raise_error(Mongo::ConnectionFailure)
      end

      it "should not raise an error when only one failed request occurs" do
        Bunyan::Logger.connection.conn_fail = 1

        lambda {
          Bunyan::Logger.count
        }.should_not raise_error(Mongo::ConnectionFailure)
      end
    end

    it "should raise errors at the mongo level" do
      # This is really a test to make sure our crazy stubbing magic
      # above is working properly. It has nothing to do with actual
      # library behavior.
      Bunyan::Logger.connection.conn_fail = 1
      db = Bunyan::Logger.db
      coll = db.collection('no_op_collection')

      lambda {
        coll.count
      }.should raise_error(Mongo::ConnectionFailure)
    end
  end
end

describe 'when initializing the collection' do
  before do
    @conn = mock_mongo_connection
    @logger = Bunyan::Logger.instance
    @logger.stub!(:connection).and_return(@conn)
  end

  it 'should create a new capped collection if the collection does not already exist' do
    @conn.should_receive(:create_collection).with('collection_1', :capped => true, :size => 52428800)
    @conn.stub!(:collection_names).and_return([])
    Bunyan::Logger.configure do |config|
      config.database   'database_1'
      config.collection 'collection_1'
    end
  end

  it 'should not create a new collection if one already exists by that name' do
    @conn.should_not_receive(:create_collection)
    @conn.collection_names.should_receive(:include?).with('collection_1').and_return(true)
    Bunyan::Logger.configure do |config|
      config.database   'database_1'
      config.collection 'collection_1'
    end
  end
end

describe 'the required config options' do
  it 'should raise an error if a db name is not provided' do
    lambda {
      Bunyan::Logger.configure do |c|
        c.collection 'collection_without_database'
      end
    }.should raise_exception(Bunyan::Logger::InvalidConfigurationError, 'Error! Please provide a database name.')
  end

  it 'should raise an error if a db collection is not provided' do
    lambda {
      Bunyan::Logger.configure do |c|
        c.database 'db_without_collection'
      end
    }.should raise_exception(Bunyan::Logger::InvalidConfigurationError, 'Error! Please provide a collection name.')
  end
end


describe Bunyan::Logger, "#disabled?" do
  it "should return false if nothing is set" do
    Bunyan::Logger.configure do |config|
      config.database   'my_database'
      config.collection 'my_collection'
    end
    Bunyan::Logger.disabled?.should == false
  end
end

describe 'mongodb instance methods passed to a logger instance' do
  it 'should be passed through to the collection' do
    configure_test_db
    Bunyan::Logger.collection.should_receive(:count)
    Bunyan::Logger.count
  end
end

describe 'alternate configuration syntax' do
  it 'should allow a user to set config options with traditional setters' do
    Bunyan::Logger.configure do |config|
      config.database = 'some_database'
      config.collection = 'some_collection'
    end
    Bunyan::Logger.config[:database].should == 'some_database'
  end
end

describe 'when bunyan is disabled' do
  it "should not send messages to the mongo collection" do
    @conn = mock_mongo_connection
    Bunyan::Logger.configure do |config|
      config.database   'bunyan_test'
      config.collection 'bunyan_test_log'
      config.disabled   true
    end
    %w(insert count find).each do |command|
      Bunyan::Logger.collection.should_not_receive(command)
      Bunyan::Logger.send command
    end
  end
end

# @NOTE: Some of these tests rely on other tests not modifying the mongo
# config object. This is not good, though much harder to achieve per-example
# independence when testing a singleton object.
describe 'when bunyan is not configured' do
  it 'should not try to send messages to mongo' do
    Bunyan::Logger.instance.stub!(:config).and_return(nil)
    %w(insert count find).each do |command|
      Bunyan::Logger.collection.should_not_receive(command)
      Bunyan::Logger.send command
    end
  end
end




spec/config_spec.rb
require 'spec_helper'

describe Bunyan::Logger::Config do
  before do
    @config = Bunyan::Logger::Config.new
  end

  it 'should have a collection' do
    @config.should respond_to :collection
  end
end

describe Bunyan::Logger::Config, 'collection size' do
  it 'should default to 50 megabytes' do
    # configure_test_db
    config = Bunyan::Logger::Config.new
    config.size.should == 52428800
  end

  it 'should all the user to set the collection size' do
    Bunyan::Logger.configure do |c|
      c.database   'bunyan_test_log'
      c.collection 'configured_size'
      c.size       100_000_000
    end

    Bunyan::Logger.config[:size].should == 100_000_000
  end
end

describe Bunyan::Logger::Config, 'when setting the collection size' do
  it 'should never set the size to nil' do
    @config = Bunyan::Logger::Config.new
    @config.size nil
    @config.size.should == 52428800
  end

  it 'should override the default value' do
    @config = Bunyan::Logger::Config.new
    @config.size 1010
    @config.size.should == 1010
  end
end

describe Bunyan::Logger::Config, 'when getting the collection size' do
  it 'should return the collection size' do
    Bunyan::Logger.configure do |c|
      c.database   'test_db'
      c.collection 'test_collection'
      c.size       2929
    end

    Bunyan::Logger.config.size.should == 2929
  end
end

describe Bunyan::Logger::Config, 'alternate method invocation syntax' do
  it 'should act like a hash' do
    config = Bunyan::Logger::Config.new
    config.size 10
    config.should_receive(:size).and_return(10)
    config[:size].should == 10
  end
end

describe 'bunyan logger configuration' do
  describe 'setting config values' do
    before do
      Bunyan::Logger.configure do |c|
        c.database   'database2'
        c.collection 'collection2'
      end
    end

    it 'should allow setting of the database' do
      Bunyan::Logger.config.database.should == 'database2'
    end

    it 'shoudl allow setting of the collection name' do
      Bunyan::Logger.config.collection.should == 'collection2'
    end
  end

  describe 'the optional config options' do
    it 'should allow the user to mark bunyan as disabled' do
      Bunyan::Logger.configure do |c|
        c.database   'test_db'
        c.collection 'test_collection'
        c.disabled   true
      end
      Bunyan::Logger.should be_disabled
    end
  end

  describe "when the disabled flag is set" do
    it 'should not create a new logger instance' do
      Bunyan::Logger.should_not_receive(:initialize_connection)
      Bunyan::Logger.configure do |c|
        c.database   'test_db'
        c.collection 'test_collection'
        c.disabled   true
      end
    end
  end
end

describe Bunyan::Logger::Config, "#disabled?" do
  before :each do
    @config = Bunyan::Logger::Config.new
  end

  it 'should be false if @disabled is nil' do
    @config.disabled = nil
    @config.should_not be_disabled
  end

  it 'should be true if @disabled is true' do
    @config.disabled = true
    @config.should be_disabled
  end

  it 'should be false if @isabled is false' do
    @config.disabled = false
    @config.should_not be_disabled
  end
end

describe 'when we configure a remote host' do
  def configure_with_remote_host
    Bunyan::Logger.configure do |c|
      c.host       'some.remote.host'
      c.database   'test_db'
      c.collection 'test_collection'
    end
  end

  it 'should attempt to connect to the remote host' do
    Mongo::Connection.should_receive(:new).with('some.remote.host', nil)
    configure_with_remote_host
  end

  it 'should set the host config option' do
    configure_with_remote_host
    Bunyan::Logger.config.host.should == 'some.remote.host'
  end
end

describe 'when we configure a port' do
  def configure_with_remote_port
    Bunyan::Logger.configure do |c|
      c.host       'some.remote.host'
      c.port       '20910'
      c.database   'test_db'
      c.collection 'test_collection'
    end
  end

  it 'should attempt to connect to a remote port' do
    Mongo::Connection.should_receive(:new).with('some.remote.host', '20910')
    configure_with_remote_port
  end

  it 'should set the port config option' do
    configure_with_remote_port
    Bunyan::Logger.config.port.should == '20910'
  end
end

describe 'when we set a connection' do

  def configure_with_connection
    Bunyan::Logger.configure do |c|
      c.connection = other_fake_mongo
      c.port       '20910'
      c.database   'test_db'
      c.collection 'test_collection'
    end
  end

  it 'should not set a connection based on other params' do
    Mongo::Connection.should_not_receive(:new)
    configure_with_connection
  end

  it 'should set the db field' do
    Mongo::Connection.should_not_receive(:new)
    configure_with_connection
    Bunyan::Logger.db.should eql(@mock_database)
  end
  
end

describe 'the new configuration syntax' do
  it 'should allow argument-less block configuration syntax' do
    Bunyan::Logger.configure do
      database   'new_config_db'
      collection 'new_config_collection'
    end
    
    Bunyan::Logger.config.database.should == 'new_config_db'
  end
end

describe 'the old configuration syntax' do
  it 'should allow the configuration object to be passed in to the configure block' do
    Bunyan::Logger.configure do |c|
      c.database   'old_config_db'
      c.collection 'old_config_collection'
    end
    
    Bunyan::Logger.config.database.should == 'old_config_db'
  end
end

describe Bunyan::Logger::Config, '#abort_on_failed_reconnect' do
  subject { Bunyan::Logger.config.abort_on_failed_reconnect }

  it 'defaults to false' do
    configure_test_db
    subject.should be_false
  end

  it 'can be set to true in the config block' do
    configure_test_db do |c|
      c.abort_on_failed_reconnect true
    end

    subject.should be_true
  end

  it 'can be set to false in the config block' do
    configure_test_db do |c|
      c.abort_on_failed_reconnect false
    end

    subject.should be_false
  end

  describe "#abort_on_failed_reconnect?" do
    subject { Bunyan::Logger.config.abort_on_failed_reconnect? }

    it "should returns true when set to true" do
      configure_test_db { |c| c.abort_on_failed_reconnect true }

      subject.should be_true
    end

    it "should return false when set to false" do
      configure_test_db { |c| c.abort_on_failed_reconnect false }
      subject.should be_false
    end

    it "should return false by default" do
      subject.should be_false
    end
  end
end



spec/spec_helper.rb
require 'rubygems'
require 'bundler'
Bundler.setup
Bundler.require :default, :test

require File.expand_path(File.dirname(__FILE__) + '/../lib/bunyan')
$LOAD_PATH.unshift File.expand_path(File.dirname(__FILE__))

class DuckedMongo
end

Spec::Runner.configure do |config|

  config.before :each do
    mock_mongo_connection
  end

  def mock_mongo_connection
    @mock_collection = mock("Mongo Collection")
    @mock_connection = mock("Mongo Connection")
    @mock_database   = mock("Mongo Database", 
       :collection        => @mock_collection, 
       :create_collection => @mock_collection,
       :collection_names  => ['name 1'],
       :connection        => @mock_connection)
    @mock_connection.stub!(:db).and_return(@mock_database)
    Mongo::Connection.stub!(:new).and_return(@mock_connection)
    @mock_database
  end

  def unstub_mongo
    Mongo::Connection.unstub!(:new)
  end

  def other_fake_mongo
    #I want to check Mongo::Connection.stub.new is not called
    dm = DuckedMongo
    dm.stub!(:db).and_return(@mock_database)
    dm
  end

  def configure_test_db(&block)
    Bunyan::Logger.configure do |c|
      c.database   'bunyan_test'
      c.collection 'bunyan_test_log'
      yield(c) if block_given?
    end
  end
end


examples/middleware.rb
# This sample middleware is provided purerly as an example. 
# It is meant to be used with authlogic.
# Please use it as your own risk.
# Once I implement some type of bunyan middleware, this file will be removed.
# - @ajsharp
class BunyanMiddleware
  def initialize(app)
    @app = app
  end
  
  # there are a number of conditions where we want to bypass and not log anything
  def call(env)
    @status, @headers, @response = @app.call(env)
    if @status != 304 && @response && !@response.body.is_a?(Proc)
      Bunyan::Logger.insert(prepare_extra_fields) 
    end
    [@status, @headers, @response]
  end

  protected
    def prepare_extra_fields
      prepare_additional_response_data.merge(prepare_user_data) || {}
    end

    def prepare_additional_response_data
      unless @response.blank?
        { 'request_method' => @response.request.request_method,
          'user_agent'     => @response.request.user_agent,
          'status'         => @status,
          'request_uri'    => @response.request.request_uri,
          'request_time'   => Time.now.utc,
          'controller_action' => format_controller_action } 
      end || {}
    end

    def format_controller_action
      params = @response.request.path_parameters
      "#{params['controller']}##{params['action']}"
    end

    def prepare_user_data
      if user_exists_in_session?
        begin
          user = User.find(@response.session[user_credentials_key])
          { 'user' => { 'email'      => user.email, 
                        'first_name' => user.first_name,
                        'last_name'  => user.last_name,
                        'roles'      => user.role_names
                      } 
          }
        rescue ActiveRecord::RecordNotFound
          {}
        end
      else
        {}
      end
    end

    def user_exists_in_session?
      !!(@response != [] && @response.session && @response.session[user_credentials_key])
    end

    def user_credentials_key
      'user_credentials_id'
    end

end


lib/bunyan.rb
require 'rubygems'
require 'mongo'
require 'singleton'

$LOAD_PATH.unshift(File.expand_path(File.dirname(__FILE__)))

require 'bunyan/config'

module Bunyan
  class Logger
    include Singleton

    class InvalidConfigurationError < RuntimeError; end

    attr_reader :db, :connection, :collection, :config

    # @example Configuring bunyan
    #   Bunyan::Logger.configure do
    #     # required options
    #     database   'bunyan_logger'
    #     collection 'development_log'
    #
    #     # optional options
    #     disabled true
    #     size 52428800 # 50.megabytes in Rails
    #   end
    def configure(&block)
      @config = Logger::Config.new
      @config.abort_on_failed_reconnect = false

      # provide legacy support for old configuration syntax
      (block.arity > 0) ? yield(@config) : @config.instance_eval(&block)

      ensure_required_options_exist
      initialize_connection unless disabled?
      @configured = true
    end

    def configured?
      !!@configured
    end

    def disabled?
      # @TODO: Refactor this. Yuck.
      config.nil? || (!config.nil? && config.disabled?)
    end

    def method_missing(method, *args, &block)
      begin
        collection.send(method, *args) if database_is_usable?
      rescue Mongo::ConnectionFailure
        # At this point, the problem may be that the server was restarted
        # and we have stale connection object. The mongo ruby driver will
        # handle automatically handling a reconnect, and will issue a fresh
        # connection object if it can obtain one. In which case, let's try
        # the query again.
        begin
          collection.send(method, *args, &block) if database_is_usable?
        rescue Mongo::ConnectionFailure => e
          # Ok, we're having real connection issues. The mongod server is likely
          # down. We still may want to fail silently, because bunyan is mostly a support
          # library, and we wouldn't want exceptions to bubble up just b/c the
          # mongod server is down. If it were the core datastore, then we probably
          # would want it to bubble up.
          #
          # If you for some reason you do want error to bubble up, set the
          # `abort_on_failed_reconnect` config option to true.

          raise e if config.abort_on_failed_reconnect?
        end
      end
    end

    # Pass all missing class methods to the singleton instance
    def self.method_missing(method, *args, &block)
      Logger.instance.send(method, *args, &block)
    end

    private
      def initialize_connection
        begin
          if (config.connection)
            @connection = config.connection
            @db = @connection.db
          else
            @db = Mongo::Connection.new(config.host, config.port).db(config.database)
            @connection = @db.connection
          end
          @collection = retrieve_or_initialize_collection(config.collection)
        rescue Mongo::ConnectionFailure => ex
          # @TODO: I don't like how I'm overloading @config.disabled
          # for user disabling and error disabling
          @config.disabled = true
          $stderr.puts 'An error occured trying to connect to MongoDB!'
        end
      end

      def database_is_usable?
        configured? && !disabled?
      end

      def ensure_required_options_exist
        raise InvalidConfigurationError, 'Error! Please provide a database name.'   unless config.database
        raise InvalidConfigurationError, 'Error! Please provide a collection name.' unless config.collection
      end

      def retrieve_or_initialize_collection(collection_name)
        if collection_exists?(collection_name)
          db.collection(collection_name)
        else
          db.create_collection(collection_name, :capped => true, :size => config.size)
        end
      end

      def collection_exists?(collection_name)
        db.collection_names.include? collection_name
      end

  end
end
